app/coffee/TikzManager.coffee

```diff
@@ -0,0 +1,37 @@
+fs = require "fs"
+Path = require "path"
+logger = require "logger-sharelatex"
+
+# for \tikzexternalize to work the main file needs to match the
+# jobname.  Since we set the -jobname to output, we have to create a
+# copy of the main file as 'output.tex'.
+
+module.exports = TikzManager =
+	needsOutputFile: (rootResourcePath, resources) ->
+		# if there's already an output.tex file, we don't want to touch it
+		for resource in resources
+			if resource.path is "output.tex"
+				return false
+		# if there's no output.tex, see if we are using tikz/pgf in the main file
+		for resource in resources
+			if resource.path is rootResourcePath
+				return TikzManager._includesTikz (resource)
+		# otherwise false
+		return false
+
+	_includesTikz: (resource) ->
+		# check if we are loading tikz or pgf
+		content = resource.content.slice(0,4096)
+		if content.indexOf("\\usepackage{tikz") >= 0
+			return true
+		else if content.indexOf("\\usepackage{pgf") >= 0
+			return true
+		else
+			return false
+
+	injectOutputFile: (compileDir, mainFile, callback = (error) ->) ->
```

Comment:
Ask discussed, refactor path validation logic in `ResourceWriter._writeResourceToDisk` into a method that can also be called here.